pl_bolts/models/self_supervised/byol/byol_module.py

```diff
@@ -75,77 +74,70 @@ def __init__(
-        return y
+    def forward(self, x: Tensor) -> Tensor:
+        """Returns the encoded representation of a view.
 
-    def shared_step(self, batch, batch_idx):
-        imgs, y = batch
-        img_1, img_2 = imgs[:2]
+        Args:
+            x (Tensor): sample to be encoded
+        """
+        return self.online_network.encode(x)
 
-        # Image 1 to image 2 loss
-        y1, z1, h1 = self.online_network(img_1)
-        with torch.no_grad():
-            y2, z2, h2 = self.target_network(img_2)
-        loss_a = -2 * F.cosine_similarity(h1, z2).mean()
+    def calculate_loss(self, v_online: Tensor, v_target: Tensor) -> Tensor:
+        """Calculates similarity loss between the online network prediction of target network projection.
 
-        # Image 2 to image 1 loss
-        y1, z1, h1 = self.online_network(img_2)
+        Args:
+            v_online (Tensor): Online network view
+            v_target (Tensor): Target network view
+        """
+        _, z1 = self.online_network(v_online)
+        h1 = self.predictor(z1)
         with torch.no_grad():
-            y2, z2, h2 = self.target_network(img_1)
-        # L2 normalize
-        loss_b = -2 * F.cosine_similarity(h1, z2).mean()
-
-        # Final loss
-        total_loss = loss_a + loss_b
-
-        return loss_a, loss_b, total_loss
-
-    def training_step(self, batch, batch_idx):
-        loss_a, loss_b, total_loss = self.shared_step(batch, batch_idx)
+            _, z2 = self.target_network(v_target)
+        loss = -2 * F.cosine_similarity(h1, z2).mean()
+        return loss
 
-        # log results
-        self.log_dict({"1_2_loss": loss_a, "2_1_loss": loss_b, "train_loss": total_loss})
+    def training_step(self, batch: Any, batch_idx: int) -> Tensor:
+        loss_12, loss_21, total_loss = self._shared_step(batch, batch_idx)
+        self.log_dict({"loss_12": loss_12, "loss_21": loss_21, "train_loss": total_loss})
+        return total_loss
 
+    def validation_step(self, batch: Any, batch_idx: int) -> Tensor:
+        loss_12, loss_21, total_loss = self._shared_step(batch, batch_idx)
+        self.log_dict({"loss_12": loss_12, "loss_21": loss_21, "val_loss": total_loss})
         return total_loss
 
-    def validation_step(self, batch, batch_idx):
-        loss_a, loss_b, total_loss = self.shared_step(batch, batch_idx)
+    def _shared_step(self, batch: Any, batch_idx: int) -> Tuple[Tensor, Tensor, Tensor]:
+        """Shared evaluation step for training and validation loop."""
+        imgs, _ = batch
+        img1, img2 = imgs[:2]
 
-        # log results
-        self.log_dict({"1_2_loss": loss_a, "2_1_loss": loss_b, "val_loss": total_loss})
+        # Calculate similarity loss in each direction
+        loss_12 = self.calculate_loss(img1, img2)
+        loss_21 = self.calculate_loss(img2, img1)
 
-        return total_loss
+        # Calculate total loss
+        total_loss = loss_12 + loss_21
+
+        return loss_12, loss_21, total_loss
```

Comment:
Can the shared step also include logging? Although `training_step` and `validation_step` are only three lines of code, essentially all of it is the same except one string, so it could really be only one line of code with a parameter